tests/test_grammar.py

```diff
@@ -137,6 +137,23 @@
     ("str_quoted_too_many_1", "''a'", GrammarParseError),
     ("str_quoted_too_many_2", "'a''", GrammarParseError),
     ("str_quoted_too_many_3", "''a''", GrammarParseError),
+    ("str_quoted_trailing_esc_1", "'abc\\\\'", "abc\\"),
+    ("str_quoted_trailing_esc_2", "'abc\\\\\\\\'", "abc\\\\"),
+    ("str_quoted_no_esc_1", '"abc\\def"', "abc\\def"),
```

Comment:
As far as I understand, this is interpreted as `abc\def` in the current grammar, right?
I understand this can simplify things, but I think there is value in simple and consistent handling of backslash escaping.

Reply:
Are you saying that *all* backslashes should be escaped? We could do that, but combined with the escaping in Python strings and the current "double un-escaping" issue (#615), this could make things quite ugly:
```python
OmegaConf.register_new_resolver("listdir", os.listdir)
cfg = OmegaConf.create({
    "all_files": "${listdir: '${drive}:\\\\\\\\path\\\\\\\\to\\\\\\\\folder'}",
    "drive": "C",
})
cfg.all_files  # will list content of C:\path\to\folder
```

Reply:
Isn't that the case now in Hydra?

The escaping in Python strings is not our concern. A user can use raw string to make things cleaner.
The double unescaping is an implementation detail, I am not sure it's really needed to make things work.

With those two statements, the above should become which seems reasonable to me.
```python
cfg = OmegaConf.create({
    "all_files": "${listdir: r'${drive}:\\path\\to\\folder'}",
    "drive": "C",
})
```

Reply:
> Isn't that the case now in Hydra? (and for that matter, in OmegaConf as well?)

Not sure about Hydra. But in OmegaConf a single backslash is ok: the fact that we un-escape double `\\` into `\` doesn't mean we have to touch single `\`.

> The escaping in Python strings is not our concern. A user can use raw string to make things cleaner.

Maybe. There are a few places where raw strings don't work though, and there's also shell escaping (e.g. with Hydra).

> The double unescaping is an implementation detail, I am not sure it's really needed to make things work.

I'm not sure either but at this point I've given up on fixing it (besides going the other way that you're suggesting, i.e. what I did in #621 where we *only* escape backslashes before quotes or interpolations).
If you have an idea that you believe may work, let me know what it is and I can try to either implement it or show you why it doesn't work :)

> With those two statements, the above should become which seems reasonable to me. (...)

Just a note that I fixed your example (the `r` was misplaced) -- it doesn't change your point.

I think that forcing double `\\` should be part of a different PR, if that's what we want to do.

Reply:
> > Isn't that the case now in Hydra? (and for that matter, in OmegaConf as well?)
> 
> Not sure about Hydra. But in OmegaConf a single backslash is ok: the fact that we un-escape double `\\` into `\` doesn't mean we have to touch single `\`.

Ok.
 
> > The escaping in Python strings is not our concern. A user can use raw string to make things cleaner.
> 
> Maybe. There are a few places where raw strings don't work though, and there's also shell escaping (e.g. with Hydra).

Escaping is not a thing in bash, They do have a syntax for literal strings though:
https://unix.stackexchange.com/questions/30903/how-to-escape-quotes-in-shell

> > The double unescaping is an implementation detail, I am not sure it's really needed to make things work.
> 
> I'm not sure either but at this point I've given up on fixing it (besides going the other way that you're suggesting, i.e. what I did in #621 where we _only_ escape backslashes before quotes or interpolations).
> If you have an idea that you believe may work, let me know what it is and I can try to either implement it or show you why it doesn't work :)

Can you create a small example and explain how it leads to double unescaping?
This will help me play with it.

> > With those two statements, the above should become which seems reasonable to me. (...)
> 
> Just a note that I fixed your example (the `r` was misplaced) -- it doesn't change your point.
> 
> I think that forcing double `\\` should be part of a different PR, if that's what we want to do.
> 
> Edit: as a minor note, this would be a breaking change to existing string interpolations in 2.0: `{"path": "${drive}:\test"}`, as I'm assuming that if we consistently enforce `\\` in quoted strings, we should also do it in top-level strings (i.e. outside of interpolations)

I don't think we should make big changes (if any) outside of quoted strings and I think I am convinced we should not require escaping of free backslash if we can make it work and be consistent.

I introduced escaping in the Hydra grammar to support single quote in a single quoted string and double quote in a double quoted string.
The introduction of nested interpolations added more use cases to escaping:

1. Nesting of quoted strings for the purposes of nested interpolations.
This is a pretty rare one. and in most cases using a different quote style should be enough.
```
'hello ${identity:"James Bond"}'
# however this turns weird if you have another level:
'hello ${identity:"James ${identity:"bond"}"}'
# or this?
'hello ${identity:"James ${identity:'bond'}"}'
# Both are wrong in their own way.

# But this should probably work:
'hello ${identity:"James ${identity:\'bond\'}"}'
```
The above does not require additional double escaping because we are using the other quote style as well.
If we didn't:
```
'hello ${identity:\'James ${identity:\\'bond\\'}'\}'
```

2. Escaping the interpolation itself:
```
\${aa}
\${foo.bar}
\${oc.env:USER}
```
multiple escaping should never be needed here. I think it's required because of your current implementation of the unescaping mechanism.

```
# no escaping:
${foo:${bar:zoo}}
# both
\${foo:\${bar:zoo}}
# inner only:
${foo:\${bar:zoo}}
```

This can be potentially be handled by introducing a token like ESC_INTER_OPEN: `\${`
By the way, this should also work outside of quoted strings because support interpolations in unquoted strings.

PS: None of the examples above are tested so mistakes are likely.

Reply:
> Escaping is not a thing in bash, They do have a syntax for literal strings though:
> https://unix.stackexchange.com/questions/30903/how-to-escape-quotes-in-shell

Really? In my zsh `echo \\` only displays one `\`. But I haven't dug deeper. Not a big deal.

> Can you create a small example and explain how it leads to double unescaping?
> This will help me play with it.

In the following example, as of the current HEAD in this branch (cd17736):

```python
from omegaconf import OmegaConf

OmegaConf.clear_resolvers()
OmegaConf.register_new_resolver("identity", lambda x: x)

cfg = OmegaConf.create({
    "x": r"${identity:'I want two \\\\\\\\ and an escaped \\\\\\${interpolation}'}",
    "interpolation": "do_not_use"
})
print(cfg.x)  # I want two \\ and an escaped \${interpolation}
```
This example prints the desired output. The goal is to cut down the number of `\` (NB: although you could use a few less `\` in the commit I mentioned, I'm assuming that we want to forbid single `\` during an un-escaping phase, so that's the correct number of `\` under this constraint).

What happens is that when we see the quoted string provided as input to the `identity` resolver (i.e., `'I want two \\\\\\\\ and an escaped \\\\\\${interpolation}'`), we do a first un-escaping pass for escaped quotes, which leads (after removing the enclosing quotes) to the string `I want two \\\\ and an escaped \\\${interpolation}`.

This string is then parsed like a regular string describing any other config option, where we do a second un-escaping pass, this time for escaped interpolations. Then it becomes `I want two \\ and an escaped \${interpolation}` which is the desired output.

Removing the double un-escaping should allow (IMO) to get the same output with:
```python
    "x": r"${identity:'I want two \\\\ and an escaped \\\${interpolation}'}",
```

With #621 this gives the desired otuput:
```python
    "x": r"${identity:'I want two \\ and an escaped \\\${interpolation}'}",
```

> 1. Nesting of quoted strings for the purposes of nested interpolations.
>    This is a pretty rare one. and in most cases using a different quote style should be enough.

Yes.

> 'hello ${identity:"James ${identity:'bond'}"}'

Just a note that this one could work. You might remember it was my first approach, so as to reduce the need for escaping. It was based on lexer modes. Not saying we should bring it back.

> 2. Escaping the interpolation itself:
> (...)
> multiple escaping should never be needed here. I think it's required because of your current implementation of the unescaping mechanism.

Based on your examples, I'm not sure we're on the same page here. The "multiple escaping" I have in mind related to escaped interpolations is when you want one (or more) slash before an interpolation: you can't use `C:\${folder}` since it would be interpreted as an escaped interpolation and lead to `C:${folder}`. So instead we need to escape the `\` and write it `C:\\${folder}`, which leads to `C:\<content_of_the_folder_node>`.

> By the way, this should also work outside of quoted strings because support interpolations in unquoted strings.

Escaped interpolations aren't supported in unquoted strings. It's not actually required since you can use a quoted string if you want an escaped interpolation. I don't think this use case is common enough to bother supporting it in unquoted strings.

Reply:
> > Can you create a small example and explain how it leads to double unescaping?
> > This will help me play with it.
> 
> In the following example, as of the current HEAD in this branch ([cd17736](https://github.com/omry/omegaconf/commit/cd17736ff7493c6f12f05329e8c11f2ab61723ad)):
> 
> ```python
> from omegaconf import OmegaConf
...
> ```

Do we want this to work? 
Can we make it work?
```python
    "x": r"${identity:'I want two \\ and an escaped \${interpolation}'}",
    "y": ${identity:${x}}, # should be the same as x
```
If it doesn't make sense to you, why?
I think it's good to focus on the desired behavior, calling out as many gotchas as we can before proposing changes.

> > 'hello ${identity:"James ${identity:'bond'}"}'
> 
> Just a note that this one could work. You might remember it was my first approach, so as to reduce the need for escaping. It was based on lexer modes. Not saying we should bring it back.

That would be nice to support, but I think we should avoid introducing "new features if possible. We can revisit for a future version.

> > 1. Escaping the interpolation itself:
> >    (...)
> >    multiple escaping should never be needed here. I think it's required because of your current implementation of the unescaping mechanism.
> 
> Based on your examples, I'm not sure we're on the same page here. The "multiple escaping" I have in mind related to escaped interpolations is when you want one (or more) slash before an interpolation: you can't use `C:\${folder}` since it would be interpreted as an escaped interpolation and lead to `C:${folder}`. So instead we need to escape the `\` and write it `C:\\${folder}`, which leads to `C:\<content_of_the_folder_node>`.

I agree with you that double escaping is needed to support this specific 

> > By the way, this should also work outside of quoted strings because support interpolations in unquoted strings.
> 
> Escaped interpolations aren't supported in unquoted strings. It's not actually required since you can use a quoted string if you want an escaped interpolation. I don't think this use case is common enough to bother supporting it in unquoted strings.

That's fine.

Reply:
> Do we want this to work?
> Can we make it work?
> 
> ```python
>     "x": r"${identity:'I want two \\ and an escaped \${interpolation}'}",
>     "y": ${identity:${x}}, # should be the same as x
> ```

I'm assuming the question is related to your example (and not my example that you quoted).

Personally, I would answer "yes" to both questions, though "work" is ambiguous since you didn't tell what the expected result should be. With #621 printing either `x` or `y` prints:
```
I want two \\ and an escaped ${interpolation}
```
which I believe can be a sensible behavior.

Just note that in my own example, I wanted to also print a `\` in front of `${interpolation}`, so if that's the goal your suggested syntax can't work (since there is a single `\` in front of `${..` it can either be printed or be used to escape the interpolation, but not both).

Reply:
## A note on Hydra

I was wondering how un-escaping worked in Hydra's quoted strings. From what I can tell, only `\"` is un-escaped into `"` (for a quoted string enclosed by double quotes). Other `\` characters aren't un-escaped.

This has the advantage of being simple and working almost everywhere... but it relies on the same regex that may not always properly detect the end of a quoted string (I haven't tried to craft a buggy example similar to #617 but I assume it should be possible). I can't think of a way to reliably detect the end of a quoted string without also requiring escaping the `\\` that may come before a quote (i.e., turning `\\\"` into `\"` when inside a quoted string).

Reply:
> > Do we want this to work?
> > Can we make it work?
> > ```python
> >     "x": r"${identity:'I want two \\ and an escaped \${interpolation}'}",
> >     "y": ${identity:${x}}, # should be the same as x
> > ```
> 
> I'm assuming your questions are related to your example (and not my example that you quoted).

Yes, somehow my very long comment got mangled, not sure how.
I meant my example you quoted.

> Personally, I would answer "yes" to both questions, though "work" is ambiguous since you didn't tell what the expected result should be. With #621 printing either `x` or `y` prints:
> 
> ```
> I want two \\ and an escaped ${interpolation}
> ```
> 
> which I believe can be a sensible behavior.

This is the interpretation I have in mind as well.

> Just note that in my own example, I wanted to also print a `\` in front of `${interpolation}`, so if that's the goal your suggested syntax can't work (since there is a single `\` in front of `${..` it can either be printed or be used to escape the interpolation, but not both).

Yes, if we want to print an escaped interpolation the input should be:
```
\\${interpolation}  # \${interpolation}
```

Another thing, the following example needs to escape the open and close curly braces in order to work:
```python
OmegaConf.register_new_resolver("str.format", str.format, replace=True)
cfg = OmegaConf.create("x: ${str.format:\{\}, 3.14}")
cfg.x # '3.14'
```
This is a bit surprising to me, those are without a $ so they should not be confused with an interpolation.
Am I missing anything?

Reply:
> ## A note on Hydra
> I was wondering how un-escaping worked in Hydra's quoted strings. From what I can tell, only `\"` is un-escaped into `"` (for a quoted string enclosed by double quotes). Other `\` characters aren't un-escaped.
> 
> This has the advantage of being simple and working almost everywhere... but it relies on the same regex that may not always properly detect the end of a quoted string (I haven't tried to craft a buggy example similar to #617 but I assume it should be possible*). I can't think of a way to reliably detect the end of a quoted string without also requiring escaping the `\\` that may come before a quote (i.e., turning `\\\"` into `\"` when inside a quoted string).
> 
> (*) Edit: actually this may not be an issue in Hydra because currently I don't think you can have anything else outside of quotes when there is a quoted string, so it may not be possible to have ambiguous situations like #617

I suspect once we start to support nested interpolations in the overrides syntax we may have to use a similar grammar to handle the similar similar situations (if we decide that we actually want to concern Hydra with the details of interpolations at all. now that OmegaConf has a proper parser generator, we might be able to simplify the override grammar and make it completely unaware of interpolations. 
We can discuss it once the dust settles on the OmegaConf grammar.

Reply:
> > Just note that in my own example, I wanted to also print a `\` in front of `${interpolation}`, so if that's the goal your suggested syntax can't work (since there is a single `\` in front of `${..` it can either be printed or be used to escape the interpolation, but not both).
> 
> Yes, if we want to print an escaped interpolation the input should be:
> 
> ```
> \\${interpolation}  # \${interpolation}
> ```

Actually with `\\${interpolation}` I would expect the result to be:
```\<content_of_the_interpolation_node_cast_to_string>```
This is similar to the `C:\\${folder}` example.

To get `\${interpolation}` you would need to use `\\\${interpolation}` (this is why the backslashes need to be escaped before an interpolation).


> 
> Another thing, the following example needs to escape the open and close curly braces in order to work:
> 
> ```python
> OmegaConf.register_new_resolver("str.format", str.format, replace=True)
> cfg = OmegaConf.create("x: ${str.format:\{\}, 3.14}")
> cfg.x # '3.14'
> ```
> 
> This is a bit surprising to me, those are without a $ so they should not be confused with an interpolation.
> Am I missing anything?

This is an unquoted string, backslashes need to be escaped to avoid confusion with a dictionary.

Reply:
> I suspect once we start to support nested interpolations in the overrides syntax we may have to use a similar grammar to handle the similar similar situations (if we decide that we actually want to concern Hydra with the details of interpolations at all. now that OmegaConf has a proper parser generator, we might be able to simplify the override grammar and make it completely unaware of interpolations.
> We can discuss it once the dust settles on the OmegaConf grammar.

Agreed on finalizing the OmegaConf grammar first (that's why I haven't touched Hydra yet).

I suspect it may be tricky though to make Hydra unaware of interpolations because detecting the end of an interpolation is far from trivial.

Reply:
Yes, it may be tricky. If we have to we can use similar grammar rules.

Reply:
> > > Just note that in my own example, I wanted to also print a `\` in front of `${interpolation}`, so if that's the goal your suggested syntax can't work (since there is a single `\` in front of `${..` it can either be printed or be used to escape the interpolation, but not both).
> > 
> > 
> > Yes, if we want to print an escaped interpolation the input should be:
> > ```
> > \\${interpolation}  # \${interpolation}
> > ```
> 
> Actually with `\\${interpolation}` I would expect the result to be:
> `\<content_of_the_interpolation_node_cast_to_string>`
> This is similar to the `C:\\${folder}` example.
> 
> To get `\${interpolation}` you would need to use `\\\${interpolation}` (this is why the backslashes need to be escaped before an interpolation).

What if we made `\${` a token for opening an escaped interpolation?
 
> This is an unquoted string, backslashes need to be escaped to avoid confusion with a dictionary.

Oh, dictionary! how quickly I evict all those details from my limited brain.

Reply:
> What if we made `\${` a token for opening an escaped interpolation?

Not sure what you mean: `\${` is already a token for escaped interpolations. This is the lexer token:
```
ESC_INTER: '\\${';
```

If you mean `\${a}` resolving to `\${a}` then how would you write a string resolving to `${a}`?


> Oh, dictionary! how quickly I evict all those details from my limited brain.

Haha no worries, honestly between quoted strings, unquoted strings, top-level strings, all having subtle differences, it's easy to lose track of such "details" :p

Reply:
> Not sure what you mean: `\${` is already a token for escaped interpolations. This is the lexer token:

Is the lexer consuming `\${` as a single token, even inside a quoted string?
I see two possible interpretations to `\\${`:
1. `\` followed by `\${a}`
2. `\\` followed by `${`

```
a: 10
b: '\\${a}'
```
b resolves to
- interpretation 1:  `\${a}`
- interpretation 2:  `\10`

Reply:
> Is the lexer consuming `\${` as a single token, even inside a quoted string?

Kind of. The lexer actually skips over the content of quoted strings, turning them into a single token `QUOTED_VALUE`.
When the grammar visitor sees a `QUOTED_VALUE`, it removes the enclosing quotes, does an un-escape pass (currently: un-escaping `\"` into `"` and `\\` into `\`, for strings with enclosing double quotes), then re-runs the whole parsing logic over the resulting string => this is when the lexer will see a `\${` as an `ESC_INTER` token (assuming it's at top level).

As example, for a quoted string `"fake \${interpolation}"`:
- Initial lexer pass: `QUOTED_VALUE` (matching `"fake \${interpolation}"`)
- After processing in the visitor: `fake \${interpolation}`
- Second lexer pass: `TOP_STR` (matching `fake `) ESC_INTER (matching `\${`) `TOP_STR` (matching `interpolation}`)

> I see two possible interpretations to `\\${`:
> 
> 1. `\` followed by `\${a}`
> 2. `\\` followed by `${`
> 
> ```
> a: 10
> b: '\\${a}'
> ```
> 
> b resolves to
> 
> * interpretation 1:  `\${a}`
> * interpretation 2:  `\10`

In interpretation 1, how do you write `b` so that it resolves to `\10`?
(in interpretation 2, you can write `b: '\\\${a}` so that it resolves to `\${a}`)

Reply:
> When the grammar visitor sees a QUOTED_VALUE, it removes the enclosing quotes, does an un-escape pass (currently: un-escaping \" into " and \\ into \, for strings with enclosing double quotes), then re-runs the whole parsing logic over the resulting string => this is when the lexer will see a \${ as an ESC_INTER token (assuming it's at top level).

Are there scenarios in the new code where the second phase is happening more than once on the same characters?
(e.g, what happens in nested quoted strings)?

> In interpretation 1, how do you write b so that it resolves to \10?

No can do.
Do you have a test that ensuring that this is covered?

> (in interpretation 2, you can write b: '\\\${a}' so that it resolves to \${a})

Can you explain the process getting from `\\\${a}` to `\${a}` (in the new code).

Reply:
> > When the grammar visitor sees a QUOTED_VALUE, it removes the enclosing quotes, does an un-escape pass (currently: un-escaping " into " and \ into , for strings with enclosing double quotes), then re-runs the whole parsing logic over the resulting string => this is when the lexer will see a ${ as an ESC_INTER token (assuming it's at top level).
> 
> Are there scenarios in the new code where the second phase is happening more than once on the same characters?
> (e.g, what happens in nested quoted strings)?

I had a long reply with a detailed example and my laptop crashed on me :/ So here's the short version, let me know if unclear:

1. "The new code" is just a grammar change (it doesn't change how the visitor works)
2. Escaped interpolations `\${` are only un-escaped in top-level strings, so they get un-escaped only once. However, currently the second phase may indeed trigger multiple un-escaping of `\\` into `\` in case of nested quoted strings (this is one thing #621 was meant to avoid).

> 
> > In interpretation 1, how do you write b so that it resolves to \10?
> 
> No can do.
> Do you have a test that ensuring that this is covered?

Yes, this is this test:
```python
("str_top_esc_backslash", r"Esc: \\${str}", r"Esc: \hi"),
```

> > (in interpretation 2, you can write b: '\${a}' so that it resolves to ${a})
> 
> Can you explain the process getting from `\\\${a}` to `\${a}` (in the new code).

It's the same in current master:
1. The lexer turns `\\\${a}` into this sequence of tokens: ESC (matching `\\`), ESC_INTER (matching `\${`), TOP_STR (maching `a}`
2. This is recognized as a `toplevelStr` by the parser
3. The visitor code visiting this node is simply `return self._unescape(ctx.getChildren())` which will concatenate the un-escaped string representation of these 3 tokens, i.e. `\`, `${` and `a}`, yielding the final result `\${a}`

Reply:
> > > When the grammar visitor sees a QUOTED_VALUE, it removes the enclosing quotes, does an un-escape pass (currently: un-escaping " into " and \ into , for strings with enclosing double quotes), then re-runs the whole parsing logic over the resulting string => this is when the lexer will see a ${ as an ESC_INTER token (assuming it's at top level).
> > 
> > 
> > Are there scenarios in the new code where the second phase is happening more than once on the same characters?
> > (e.g, what happens in nested quoted strings)?
> 
> I had a long reply with a detailed example and my laptop crashed on me :/ So here's the short version, let me know if unclear:
> 
> 1. "The new code" (in this PR) is just a grammar change (it doesn't change how the visitor works)
> 2. Escaped interpolations `\${` are only un-escaped in top-level strings, so they get un-escaped only once. However, currently the second phase may indeed trigger multiple un-escaping of `\\` into `\` in case of nested quoted strings (this is one thing #621 was meant to avoid).

Okay, so once this is in, #621 is supposed to address the repeated un-escaping issue?
 
> > Can you explain the process getting from `\\\${a}` to `\${a}` (in the new code).
> 
> It's the same in current master:
> 
> 1. The lexer turns `\\\${a}` into this sequence of tokens: ESC (matching `\\`), ESC_INTER (matching `\${`), TOP_STR (maching `a}`
> 2. This is recognized as a `toplevelStr` by the parser
> 3. The visitor code visiting this node is simply `return self._unescape(ctx.getChildren())` which will concatenate the un-escaped string representation of these 3 tokens, i.e. `\`, `${` and `a}`, yielding the final result `\${a}`

Gotcha, this answers my question.
Can you explain why we need a toplevelStr?

Reply:
> Okay, so once this is in, #621 is supposed to address the repeated un-escaping issue?

That was my original plan, though I'm open to other approaches for un-escaping (including not actually fixing it).

> Can you explain why we need a toplevelStr?

This is for string interpolations, so that users can use any character they want outside of the interpolation part, including quotes. You might remember an example we discussed back in the days, that was something like: `"Hi ${name}", he said`. Here, there would be two top-level strings: `"Hi ` and `", he said`.

Reply:
> That was my original plan, though I'm open to other approaches for un-escaping (including not actually fixing it).

Cool. I think we should try to fix it. I am asking all those questions because I am trying to reach the best solution, not because I think this is not important enough to fix.

> This is for string interpolations, so that users can use any character they want outside of the interpolation part, including quotes. You might remember an example we discussed back in the days, that was something like: "Hi ${name}", he said. Here, there would be two top-level strings: "Hi and ", he said.

Gotcha.